### 1️⃣ Data Retrieval and Collection

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


df = pd.read_csv('Salary_Data.csv')

print("Dataset Shape:", df.shape)
print("\nColumn Names:", df.columns.tolist())
print("\nFirst 5 Rows:")
print(df.head())
print("\nDataset Info:")
print(df.info())

Dataset Shape: (6704, 6)

Column Names: ['Age', 'Gender', 'Education Level', 'Job Title', 'Years of Experience', 'Salary']

First 5 Rows:
    Age  Gender Education Level          Job Title  Years of Experience  \
0  32.0    Male      Bachelor's  Software Engineer                  5.0   
1  28.0  Female        Master's       Data Analyst                  3.0   
2  45.0    Male             PhD     Senior Manager                 15.0   
3  36.0  Female      Bachelor's    Sales Associate                  7.0   
4  52.0    Male        Master's           Director                 20.0   

     Salary  
0   90000.0  
1   65000.0  
2  150000.0  
3   60000.0  
4  200000.0  

Dataset Info:
<class 'pandas.DataFrame'>
RangeIndex: 6704 entries, 0 to 6703
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Age                  6702 non-null   float64
 1   Gender               6702 non-null   str    
 2   Education Le

### 2️⃣ Data Cleaning

In [22]:
# Check for missing values
missing_values = df.isnull().sum()
print("Missing values per column:\n", missing_values)
df['Years of Experience'] = pd.to_numeric(df['Years of Experience'], errors='coerce')

# Remove missing values

df = df.dropna()

# Verify data types
print("\nData types:\n", df.dtypes)

Missing values per column:
 Age                    0
Gender                 0
Education Level        0
Job Title              0
Years of Experience    0
Salary                 0
dtype: int64

Data types:
 Age                    float64
Gender                     str
Education Level            str
Job Title                  str
Years of Experience    float64
Salary                 float64
dtype: object





Missing Values ​​Treatment: (Standard Salary datasets usually have 0 missing values). If any missing values ​​are detected, they should either be dropped (if negligible) or imputed using the mean/median. For this clean dataset, no imputation is required.


Data Types: Both YearsExperienceand Salaryare verified as numeric float/int values, which are perfectly compatible with mathematical regression models.

### 3️⃣ Feature Design

In [23]:


X = df[['Years of Experience']]
y = df['Salary']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print(f"Training set size: {X_train.shape[0]} samples")
print(f"Testing set size: {X_test.shape[0]} samples")

Training set size: 5358 samples
Testing set size: 1340 samples


### 4️⃣ Algorithm Selection
We select Simple Linear Regression for this task.

Why is it appropriate? 

Linear Regression is highly suitable here because our target variable ( Salary) is continuous, and we are investigating a direct, expectedly linear relationship with a single continuous predictor ( Years Of Experience). It provides a highly interpretable baseline model to observe how a change in career duration directly scales income.

### 5️⃣ Loss Function Selection

We select the Mean Squared Error (MSE) as our objective loss function, mathematically expressed as:
$$MSE = \frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2$$

What does MSE measure? 

MSE measures the average of the squares of the errors (the differences between the actual values $y_i$ and predicted values
$\hat{y}_i$). By squaring the errors, MSE heavily penalizes larger deviations or outliers, forcing the regression line to find the optimal balance that minimizes massive mispredictions.

### 6️⃣ Model Learning

In [24]:

model = LinearRegression()

model.fit(X_train, y_train)

print("Model training complete.")

Model training complete.


### 7️⃣ Model Evaluation & Interpretation

In [25]:
# Make predictions
y_pred = model.predict(X_test)

# Calculate Evaluation Metrics
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Extract Coefficients
coefficient = model.coef_[0]
intercept = model.intercept_

print(f"--- Evaluation Metrics ---")
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R² Score: {r2:.4f}")
print(f"\n--- Model Parameters ---")
print(f"Coefficient (Slope): {coefficient:.2f}")
print(f"Intercept: {intercept:.2f}")

--- Evaluation Metrics ---
Mean Squared Error (MSE): 982050061.92
R² Score: 0.6556

--- Model Parameters ---
Coefficient (Slope): 7072.86
Intercept: 58196.21


Model Interpretation:

 1. The Intercept ($58,196.21$)
 
 What it means mathematically:

  The intercept is the value of $y$ (Salary) when $X$ (YearsExperience) is exactly equal to $0$.
   It is the point where the regression line crosses the vertical y-axis.
   
   What it means in this context: 
   The intercept represents the baseline starting salary for an employee with zero years of experience . 
   According to your model, a fresh graduate or an entry-level hire entering the workforce can expect a predicted starting salary of $58,196.21 .

2. The Coefficient / Slope ($7,072.86$)

What it means mathematically: 

The coefficient represents the rate of change of the line. 

It tells you how much $y$ changes for every single-unit increase in $X$ .

What it means in this context: 
The coefficient represents the financial value of one additional year of experience . For every 1-year increase in an employee's experience, their salary is predicted to increase by a steady average of $7,072.86 .